File: integrations/python/control.py
#!/usr/bin/env python3
from __future__ import annotations
import json, os, pathlib, time, urllib.error, urllib.request

INFERENCE = "http://127.0.0.1:8080"
CONTROL = "http://127.0.0.1:18081"


def api_key() -> str:
    value = os.environ.get("FLASHNEXT_API_KEY")
    if value:
        return value
    path = pathlib.Path(os.environ["LOCALAPPDATA"]) / "FlashNextManager" / "api-key.txt"
    value = path.read_text(encoding="utf-8").strip()
    if len(value) < 32:
        raise RuntimeError("FlashNext API key is invalid.")
    return value


def request(method: str, path: str, payload: dict | None = None, timeout: float = 120, base: str = CONTROL) -> dict | str:
    data = None if payload is None else json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        base + path,
        data=data,
        headers={"Authorization": f"Bearer {api_key()}", "Content-Type": "application/json"},
        method=method,
    )
    with urllib.request.urlopen(req, timeout=timeout) as response:
        body = response.read().decode("utf-8")
        if not body:
            return {}
        try:
            return json.loads(body)
        except json.JSONDecodeError:
            return body


def wait_until_healthy(timeout_seconds: float = 900, poll_seconds: float = 2) -> None:
    """Poll GET http://127.0.0.1:8080/health until llama-server is ready. Does not start the model."""
    deadline = time.time() + timeout_seconds
    last_error = "inference /health was never reached"
    while time.time() < deadline:
        try:
            request("GET", "/health", timeout=min(5, timeout_seconds), base=INFERENCE)
            return
        except (urllib.error.URLError, urllib.error.HTTPError, TimeoutError, OSError) as ex:
            last_error = str(ex)
            time.sleep(poll_seconds)
    raise TimeoutError(f"FlashNext inference server was not healthy within {timeout_seconds:.0f}s: {last_error}")


def start_and_wait(timeout_seconds: float = 900) -> dict | str:
    """POST control start, then wait until :8080/health succeeds."""
    started = request("POST", "/flashnext/server/start", timeout=timeout_seconds)
    wait_until_healthy(timeout_seconds=timeout_seconds)
    return started


if __name__ == "__main__":
    # Sampling (temperature, thinking, max_tokens) can also be sent on each
    # POST http://127.0.0.1:8080/v1/chat/completions with no restart.
    # PATCH mtpNMax / contextSize / specType / port then POST /flashnext/server/restart.
    status = request("GET", "/flashnext/status")
    print(status)
    patched = request("PATCH", "/flashnext/settings", {"profiles": {"coding-balanced": {"temperature": 0.8}}})
    print("restartRequired", patched.get("restartRequired") if isinstance(patched, dict) else patched)

File: integrations/csharp/Control.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

// Wait for GET http://127.0.0.1:8080/health before POST /v1/chat/completions.
// Sampling fields may be sent on each chat request with no restart.
// PATCH mtpNMax / contextSize / specType / port / extraArguments, then POST flashnext/server/restart.
string status = await FlashNextClient.ControlGetAsync("flashnext/status");
Console.WriteLine(status);
using JsonDocument patched = await FlashNextClient.PatchSettingsAsync("""{"profiles":{"coding-balanced":{"temperature":0.8}}}""");
Console.WriteLine("restartRequired " + patched.RootElement.GetProperty("restartRequired").GetBoolean());

public static class FlashNextClient
{
    public static string ApiKey()
    {
        string? env = Environment.GetEnvironmentVariable("FLASHNEXT_API_KEY");
        if (!string.IsNullOrWhiteSpace(env)) return env.Trim();
        string path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "FlashNextManager", "api-key.txt");
        string value = File.ReadAllText(path).Trim();
        if (value.Length < 32) throw new InvalidDataException("FlashNext API key is invalid.");
        return value;
    }

    public static HttpClient InferenceClient(TimeSpan? timeout = null)
    {
        HttpClient client = new() { BaseAddress = new Uri("http://127.0.0.1:8080/"), Timeout = timeout ?? TimeSpan.FromMinutes(30) };
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiKey());
        return client;
    }

    public static HttpClient ControlClient(TimeSpan? timeout = null)
    {
        HttpClient client = new() { BaseAddress = new Uri("http://127.0.0.1:18081/"), Timeout = timeout ?? TimeSpan.FromMinutes(15) };
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiKey());
        return client;
    }

    public static async Task WaitUntilHealthyAsync(TimeSpan? timeout = null, CancellationToken cancellationToken = default)
    {
        TimeSpan limit = timeout ?? TimeSpan.FromSeconds(900);
        DateTimeOffset deadline = DateTimeOffset.UtcNow + limit;
        Exception? last = null;
        using HttpClient client = InferenceClient(TimeSpan.FromSeconds(5));
        while (DateTimeOffset.UtcNow < deadline)
        {
            cancellationToken.ThrowIfCancellationRequested();
            try
            {
                using HttpResponseMessage response = await client.GetAsync("/health", cancellationToken).ConfigureAwait(false);
                if (response.IsSuccessStatusCode) return;
                last = new HttpRequestException("Inference /health returned " + (int)response.StatusCode);
            }
            catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
            {
                last = ex;
            }
            await Task.Delay(TimeSpan.FromSeconds(2), cancellationToken).ConfigureAwait(false);
        }
        throw new TimeoutException("FlashNext inference server was not healthy within " + limit + ". " + last?.Message);
    }

    public static async Task StartAndWaitAsync(CancellationToken cancellationToken = default)
    {
        using HttpClient control = ControlClient();
        using HttpResponseMessage start = await control.PostAsync("flashnext/server/start", null, cancellationToken).ConfigureAwait(false);
        start.EnsureSuccessStatusCode();
        await WaitUntilHealthyAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
    }

    public static async Task<string> ControlGetAsync(string path, CancellationToken cancellationToken = default)
    {
        using HttpClient control = ControlClient();
        return await control.GetStringAsync(path, cancellationToken).ConfigureAwait(false);
    }

    public static async Task<JsonDocument> PatchSettingsAsync(string json, CancellationToken cancellationToken = default)
    {
        using HttpClient control = ControlClient();
        using HttpRequestMessage patch = new(HttpMethod.Patch, "flashnext/settings")
        {
            Content = new StringContent(json, Encoding.UTF8, "application/json")
        };
        using HttpResponseMessage response = await control.SendAsync(patch, cancellationToken).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
        return JsonDocument.Parse(await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false));
    }
}
